File: LeetCode/101.py
# Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center)

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    # iterative
    def isSymmetric1(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True

        else:
            stack = [root.left, root.right]
            while stack:
                node1 = stack.pop()
                node2 = stack.pop()

                if not node1 and not node2:
                    return True
                if not node1 or not node2:
                    return False
                if node1.val != node2.val:
                    return False

                else:
                    if node1.left and node2.right:
                        stack += [node1.left, node2.right]
                    elif node1.left or node2.right:
                        return False

                    if node2.left and node1.right:
                        stack += [node2.left, node1.right]
                    elif node2.left or node1.right:
                        return False

            return True

    # recursive
    def isSymmetric2(self, root):
        if not root:
            return True
        else:
            return self.dfs(root.left, root.right)

    def dfs(self, node1, node2):
        if not node1 and not node2:
            return True
        if not node1 or not node2:
            return False

        if node1.val != node2.val:
            return False

        return self.dfs(node1.left, node2.right) and self.dfs(node1.right, node2.left)

File: Cpp/LeetCode/101.cpp
//
// Created by caizixiang on 2019-07-30.
//

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root)
            return true;
        else
            return dfs(root->left, root->right);
    }

    bool dfs(TreeNode* node1, TreeNode* node2)
    {
        if (!node1 && !node2)
            return true;
        else if (node1 && node2)
        {
            if (node1->val != node2->val)
                return false;
            else
            {
                return dfs(node1->left, node2->right) && dfs(node1->right, node2->left);
            }
        }
        else
            return false;
    }
};